libnavui-maneuver/src/main/java/com/mapbox/navigation/ui/maneuver/ManeuverProcessor.kt

```diff
@@ -3,92 +3,290 @@ package com.mapbox.navigation.ui.maneuver
+                    legStep.maneuverList[i].stepDistance.distanceRemaining = stepDistanceRemaining
+                    break
+                }
+            }
+        }
+        return maneuverIndex
+    }
+
+    private fun filterList(
+        maneuverIndex: Int,
+        indexOfLegStep: Int,
+        inputList: List<StepIndexToManeuvers>
+    ): List<Maneuver> {
+        val list = mutableListOf<Maneuver>()
+        if (maneuverIndex == Int.MIN_VALUE) {
+            for (i in indexOfLegStep..inputList.lastIndex) {
+                list.addAll(inputList[i].maneuverList)
+            }
+        } else {
+            for (i in indexOfLegStep..inputList.lastIndex) {
+                val maneuverList = inputList[i].maneuverList
+                if (i == indexOfLegStep && maneuverList.size > 1) {
+                    list.addAll(maneuverList.subList(maneuverIndex, maneuverList.size))
+                } else {
+                    list.addAll(maneuverList)
+                }
+            }
+        }
+        return list
+    }
+
+    private fun transformToManeuver(
+        bannerInstruction: BannerInstructions,
+        distanceFormatter: DistanceFormatter,
+        routeProgress: RouteProgress? = null
+    ): Maneuver {
         val primaryManeuver = getPrimaryManeuver(bannerInstruction.primary())
         val secondaryManeuver = getSecondaryManeuver(bannerInstruction.secondary())
         val subManeuver = getSubManeuverText(bannerInstruction.sub())
-        val totalDistance = getTotalStepDistance(bannerInstruction)
         val laneGuidance = getLaneGuidance(bannerInstruction)
-        val maneuver = Maneuver(
+        val totalStepDistance = bannerInstruction.distanceAlongGeometry()
+        val stepDistanceRemaining = getStepDistanceRemaining(routeProgress)
+        val stepDistance = StepDistance(distanceFormatter, totalStepDistance, stepDistanceRemaining)
+        return Maneuver(
             primaryManeuver,
-            totalDistance,
+            stepDistance,
             secondaryManeuver,
             subManeuver,
             laneGuidance
         )
-        return ManeuverResult.GetManeuver(maneuver)
     }
 
-    private fun getTotalStepDistance(bannerInstruction: BannerInstructions) =
-        TotalManeuverDistance(bannerInstruction.distanceAlongGeometry())
+    private fun getStepDistanceRemaining(routeProgress: RouteProgress?): Double? {
+        return ifNonNull(routeProgress?.currentLegProgress?.currentStepProgress) {
+            it.distanceRemaining.toDouble()
+        }
+    }
 
-    private suspend fun getPrimaryManeuver(bannerText: BannerText): PrimaryManeuver {
+    private fun getPrimaryManeuver(bannerText: BannerText): PrimaryManeuver {
         val bannerComponentList = bannerText.components()
         return when (!bannerComponentList.isNullOrEmpty()) {
             true -> {
                 PrimaryManeuver
                     .Builder()
+                    .id(UUID.randomUUID().toString())
```

Comment:
Why do we need an `UUID`? What use case are we covering appending it to the `Maneuver`s?

Reply:
This is a big and complicated topic that is difficult to discuss here. But, to sum it up, to attain the asynchronous behavior of first fetching only the text instruction and then fetching the road shield, it is important for the API to know which maneuver to map the road shield to. Hence, an id for each `PrimaryManeuver`, `SecondaryManeuver` and `SubManeuver` helps the API to easily identify the maneuver and it's type and associate the road shield with the appropriate maneuver. This id and roadShield is then easily available to the end user who can extract the maneuver based on the id and append the road shield to the maneuver.

cc @Guardiola31337

Reply:
But the `UUID` is random 🤔 

We can discuss further internally 👍

Reply:
@Guardiola31337 the ID creates a mapping between the banner and its downloaded shield. This is only stored in memory, not on disk, so we can generate random values when a banner instance is first created in the runtime and use that as a base for the mapping.

Reply:
But ultimately that's available publicly in the `Maneuver`s. Let me rephrase the question. How's the `id` useful for end users? Isn't adding this confusing if developers can't use this? 🤔

Reply:
If developers were not using our `ManeuverView` and instead built their own, the only way to associate a shield bitmap with a banner is via this ID.

Reply:
@abhishek1508 we should add a section to the `MapboxManeuverApi` class docs that explains this relationship.